server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -77,65 +87,212 @@
 
+        runTasksUntil(deterministicTaskQueue, maxDisconnectionTime);
+
         // disable exceptions so things can be restored
         transport.randomConnectionExceptions = false;
-        service.new ConnectionChecker().run();
-        assertConnectedExactlyToNodes(event.state());
+        logger.info("renewing connections");
+        runTasksUntil(deterministicTaskQueue, maxDisconnectionTime + reconnectIntervalMillis);
+        assertConnectedExactlyToNodes(targetNodes);
+    }
+
+    public void testOnlyBlocksOnConnectionsToNewNodes() throws Exception {
+        final NodeConnectionsService service = new NodeConnectionsService(Settings.EMPTY, threadPool, transportService);
+
+        // connect to one node
+        final DiscoveryNode node0 = new DiscoveryNode("node0", buildNewFakeTransportAddress(), Version.CURRENT);
+        final DiscoveryNodes nodes0 = DiscoveryNodes.builder().add(node0).build();
+        final PlainActionFuture<Void> future0 = new PlainActionFuture<>();
+        service.connectToNodes(nodes0, () -> future0.onResponse(null));
+        future0.actionGet();
+        assertConnectedExactlyToNodes(nodes0);
+
+        // connection attempts to node0 block indefinitely
+        final CyclicBarrier connectionBarrier = new CyclicBarrier(2);
+        try {
+            nodeConnectionBlocks.put(node0, connectionBarrier::await);
+            transportService.disconnectFromNode(node0);
+
+            // can still connect to another node without blocking
+            final DiscoveryNode node1 = new DiscoveryNode("node1", buildNewFakeTransportAddress(), Version.CURRENT);
+            final DiscoveryNodes nodes1 = DiscoveryNodes.builder().add(node1).build();
+            final DiscoveryNodes nodes01 = DiscoveryNodes.builder(nodes0).add(node1).build();
+            final PlainActionFuture<Void> future1 = new PlainActionFuture<>();
+            service.connectToNodes(nodes01, () -> future1.onResponse(null));
+            future1.actionGet();
+            assertConnectedExactlyToNodes(nodes1);
+
+            // can also disconnect from node0 without blocking
+            final PlainActionFuture<Void> future2 = new PlainActionFuture<>();
+            service.connectToNodes(nodes1, () -> future2.onResponse(null));
+            future2.actionGet();
+            service.disconnectFromNodesExcept(nodes1);
+            assertConnectedExactlyToNodes(nodes1);
+
+            // however, now node0 is considered to be a new node so we will block on a subsequent attempt to connect to it
+            final PlainActionFuture<Void> future3 = new PlainActionFuture<>();
+            service.connectToNodes(nodes01, () -> future3.onResponse(null));
+            expectThrows(ElasticsearchTimeoutException.class, () -> future3.actionGet(timeValueMillis(scaledRandomIntBetween(1, 1000))));
+
+            // once the connection is unblocked we successfully connect to it.
+            nodeConnectionBlocks.clear();
+            connectionBarrier.await();
```

Comment:
`connectionBarrier.await(0, TimeUnit.SECONDS)`, ie. zero timeout to avoid waiting. I think we expect another thread to hold the barrier at this time?

Reply:
Thanks, ddaa024a2e69c322df1b7ceee84d0872024dc49b